packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js

```diff
@@ -0,0 +1,74 @@
+const path = require(`path`)
+
+module.exports = class GatsbyThemeComponentShadowingResolverPlugin {
+  cache = {}
+
+  constructor({ projectRoot, themes }) {
+    this.themes = themes
+    this.projectRoot = projectRoot
+  }
+
+  apply(resolver) {
+    resolver.plugin(`relative`, (request, callback) => {
+      // find out which theme's src/components dir we're requiring from
+      const matchingThemes = this.themes.filter(name =>
+        request.path.includes(path.join(name, `src`, `components`))
+      )
+      // 0 matching themes happens a lot fo rpaths we don't want to handle
+      // > 1 matching theme means we have a path like
+      //   `gatsby-theme-blog/src/components/gatsby-theme-something/src/components`
+      if (matchingThemes.length > 1) {
+        throw new Error(
+          `Gatsby can't differentiate between themes ${matchingThemes.join(
+            ` and `
+          )} for path ${request.path}`
+        )
+      }
+      if (matchingThemes.length !== 1) {
+        return callback()
+      }
+      // theme is the theme package from which we're requiring the relative component
+      const [theme] = matchingThemes
+      // get the location of the component relative to src/components
+      const [, component] = request.path.split(
+        path.join(theme, `src`, `components`)
```

Comment:
[gatsby-plugin-page-creator](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-page-creator) already supports being added multiple times. Since its an internal plugin, I think we should [add one for each theme](https://github.com/ChristopherBiscardi/gatsby-theme-examples/blob/89ae0815e3a20b5b506ce06c324d28669e90dad3/themes/gatsby-theme-blog/gatsby-config.js#L73-L78) if we want them to be overridable by default (I think we do). The example I just linked to shows the multiple-plugins approach implemented manually in userland.

Reply:
Okay, cool. To be clear, does that mean that if a theme includes pages, the user would need to implement both the theme and `gatsby-plugin-page-creator` to override?

I also just thought about templates.

I'm imagining the blog theme, for instance, shipping with a `src/pages/previews.js` and a `src/templates/post.js` that create defaults for a blogging theme. In my child theme, I want to change the post template, so I create my own that adheres to the original `post.js` data format — how would that work? Do I need to fully eject the `gatsby-node.js`, or could I just create `gatsby-theme-blog/src/templates/post.js` and trust the theme to override for me?

Reply:
> Okay, cool. To be clear, does that mean that if a theme includes pages, the user would need to implement both the theme and gatsby-plugin-page-creator to override?

No, the theme would need to implement gatsby-plugin-page-creator if it wanted to offer pages. The user's site always owns the internal instance of the plugin right now. (I'm considering whether we should opt themes into their own instance of the plugin by default or not in the same way we currently do for sites)

---

> how would that work? Do I need to fully eject the gatsby-node.js, or could I just create gatsby-theme-blog/src/templates/post.js and trust the theme to override for me?

I imagine a best practice will evolve where the theme's page templates more or less just contain the graphql query and the template imports a component from src/components to handle rendering the data. This conventional separation of query and rendering component seems like its going to be part of the theming world to enable users to take full advantage of progressive disclosure for shadowing.

---

I played with the idea of automatically shadow-enabling any createPage call and since you can override page templates using the gatsby lifecycle methods (onCreatePage) I chose to not implement automatic shadowing so as to not create multiple APIs to do the same thing to start. I'm lightly of the opinion that this contributes to a progressive disclosure in the onboarding process from new Gatsby theme user to theme creator. Roughly:

* User shadows a component (taking control of a specific minimal sub-tree)
* User shadows theme tokens (taking control of font size, brand colors, etc, only possible if set up by the theme)
* User shadows a query (adding more props to a component; only possible by itself if the theme author separates their query components from the render components)
* User takes control of a page (taking control of entire sub-tree, introduction to gatsby lifecycles onCreatePage)
* User adds fields to pre-existing data (gatsby lifecycle + graphql querying adjustments + rendering)
* User creates their own data (starting to mess with including their own plugins, writing their own createPage calls for that data, etc),

** creating your own pages as .js files sits somewhere near the "early effort" end of the spectrum because of plugin-page-creator.

Reply:
Alright, great — this all makes sense. I love the idea of progressive disclosure here, and it sounds like you're thinking through the story really well.

Great work!